gostream/ffmpeg/avcodec.go

```diff
@@ -0,0 +1,679 @@
+//go:build cgo && !android
+
+package ffmpeg
+
+//#include <libavformat/avformat.h>
+//#include <libavcodec/avcodec.h>
+//#include <libavcodec/packet.h>
+//#include <libavutil/avutil.h>
+//#include <libavutil/imgutils.h>
+//#include <libswscale/swscale.h>
+// static const char *error2string(int code) { return av_err2str(code); }
+import "C"
+
+import (
+	"context"
+	"image"
+	"unsafe"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	// YUV420P the pixel format AV_PIX_FMT_YUV420P
+	YUV420P = C.AV_PIX_FMT_YUV420P
+	// Target bitrate in bits per second.
+	bitrate = 1_200_000
+	// Target bitrate tolerance factor.
+	// E.g., 2.0 gives 100% deviation from the target bitrate.
+	bitrateDeviation = 2
+)
+
+// From https://ffmpeg.org/doxygen/trunk/pixfmt_8h_source.html#l00064
+const (
+	// PixFmtNone an AV_PIX_FMT_NONE
+	PixFmtNone = iota - 1
+	// PixFmtYUV420P an AV_PIX_FMT_YUV420P
+	PixFmtYUV420P
+	// PixFmtYUYV422 an AV_PIX_FMT_YUYV422
+	PixFmtYUYV422
+	// PixFmtRGB24 an AV_PIX_FMT_RGB24
+	PixFmtRGB24
+	// PixFmtBGR24 an AV_PIX_FMT_BGR24
+	PixFmtBGR24
+	// PixFmtYUV422P an AV_PIX_FMT_YUV422P
+	PixFmtYUV422P
+	// PixFmtYUV444P an AV_PIX_FMT_YUV444P
+	PixFmtYUV444P
+	// PixFmtYUV410P an AV_PIX_FMT_YUV410P
+	PixFmtYUV410P
+	// PixFmtYUV411P an AV_PIX_FMT_YUV411P
+	PixFmtYUV411P
+	// PixFmtGray8 an AV_PIX_FMT_GRAY8
+	PixFmtGray8
+	// PixFmtMonoWhite an AV_PIX_FMT_MONOWHITE
+	PixFmtMonoWhite
+	// PixFmtMonoBlack an AV_PIX_FMT_MONOBLACK
+	PixFmtMonoBlack
+	// PixFmtPal8 an AV_PIX_FMT_PAL8
+	PixFmtPal8
+	// PixFmtYUVJ420P an AV_PIX_FMT_YUVJ420P
+	PixFmtYUVJ420P
+	// PixFmtYUVJ422P an AV_PIX_FMT_YUVJ422P
+	PixFmtYUVJ422P
+	// PixFmtYUVJ444P an AV_PIX_FMT_YUVJ444P
+	PixFmtYUVJ444P
+	// PixFmtUYVY422 an AV_PIX_FMT_UYVY422
+	PixFmtUYVY422
+	// PixFmtUYYVYY411 an AV_PIX_FMT_UYYVYY411
+	PixFmtUYYVYY411
+	// PixFmtBGR8 an AV_PIX_FMT_BGR8
+	PixFmtBGR8
+)
+
+type (
+	// Codec an AVCodec
+	Codec struct {
+		p  *C.struct_AVCodec
+		pp **C.struct_AVCodec
+	}
```

Comment:
I was having issues passing double pointers to C functions. This solved the issue and made dealing with C functions with pointer and double pointer args cleaner.